### ISSUE TITLE
Narrowing a union of TypedDicts during construction

### ISSUE THREAD

**Issue by tungol** [python/mypy#15874]
**Feature**

I am not totally sure if this a bug or a feature request. When constructing a dictionary typed to be the union of two TypedDicts, I was surprised that I couldn't add a non-overlapping key.

**Pitch**

Consider:

```python
from typing import TypedDict


class A(TypedDict):
    foo: str


class B(TypedDict):
    foo: str
    bar: str


def get_thing(bar: str | None = None) -> A | B:
    result: A | B = {"foo": "hi"}
    if bar:
        result["bar"] = bar  # current behavior: error: TypedDict "A" has no key "bar"
    return result
```

In the actual code, the types are more complicated and I have reasons not to just use NotRequired.

Adding "bar" seemed like a form of narrowing the Union to me, so I was surprised that this wasn't allowed. I can't use the 'in' method of narrowing yet, for obvious reasons.

Anyway, it was easy enough to work around with a cast, so I'd understand if this is just not something that mypy wants to support.

**Comment by erictraut** [comment]
What you're asking for is not sound from a type perspective. When you create an object from the expression `{"foo": "hi"}`, that object is an `A`, not a `B`. You are allowed to assign it to a variable whose declared type is `A | B`, but that doesn't change the fact that the value you assigned is a type `A`. 

Here's an analogy that might make this clearer.

```python
x: list[str] | list[int]
x = ["a", "b", "c"] # This assignment is allowed, but x is now narrowed to `list[str]`
x.append([1, 2, 3]) # Type error
```

In my example the object needs to either "start life" as a `list[str]` or a `list[int]`, and once it is constructed, it cannot change types.

Likewise, in your example, the object starts life as an `A`, so it can't change into a `B` without constructing a new object. So this is legal:

```python
def get_thing(bar: str | None = None) -> A | B:
    result: A | B = {"foo": "hi"}
    if bar:
        result = {**result, "bar": bar}
    return result
```

**Comment by tungol** [comment]
Interesting. I suspected something like that, but `reveal_type` showed the union, which muddled the issue. The difference between this and your analogous case though, is the nature of partially overlapping TypedDicts. There's no overlap between `list[str]` and `list[int]`, whereas I can make TypedDicts that are genuinely ambiguous.

Let's make my original case more ambiguous:

```
from typing import TypedDict, NotRequired

class A(TypedDict):
    foo: str
    baz: NotRequired[str]


class B(TypedDict):
    foo: str
    bar: NotRequired[str]


def get_thing(bar: str | None = None) -> A | B:
    result: A | B = {"foo": "hi"}
    if bar:
        result["bar"] = bar
    return result
```

Is it really accurate to say that `result` "starts life" as either A or B here?

**Comment by erictraut** [comment]
Yes, it's accurate to say that `result` "starts life" as an `A`. Let's consider the implications if that weren't the case and you could modify its type after the fact.

```python
class A(TypedDict):
    foo: str
    bar: NotRequired[int]

class B(TypedDict):
    foo: str
    bar: NotRequired[str]

a: A
b: B

# If {"foo": "hi"} is ambiguous, then this should be allowed without a type violation.
a = b = {"foo": "hi"}

# But this mutation violates the type of `a`.
b["bar"] = "hi"
```

**Comment by ikonst** [comment]
FWIW, same on TypeScript: https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgILIN4Chm+TAewIC5kBnMKUAcywF8stRJZEUAhTHPQk8ymgG5uuAEZwopClRDVhDLDACuIBGGAEQyahDAB9MAAsaACnFQA-FIGyAlKXQAfZJ2x5kANwnIoEMkoAbMAdkZ04AXkxeUgAiYxiGd2AYZDMJWy53d19-IIA6c2RI82F3RLxfMCUoLRzAsHosIA

**Comment by tungol** [comment]
Makes sense. Thanks - we can probably just close this then.

**Comment by tungol** [comment]
Oh, and the suggested workaround:

```
def get_thing(bar: str | None = None) -> A | B:
    result: A | B = {"foo": "hi"}
    if bar:
        result = {**result, "bar": bar}
    return result
```
doesn't hit that particular error, but does hit various others. I got:

```
error: Incompatible types in assignment (expression has type "Dict[str, str]", variable has type ___)
error: Expected TypedDict key to be string literal
error: Argument 1 to "update" of "MutableMapping" has incompatible type __; expected "SupportsKeysAndGetItem[str, str]"
```

Whereas if I supply an explicit type on the redefinition I get:

```
error: Name "result" already defined on line 91
```

so a cast seems to be the most workable option still.

**Comment by erictraut** [comment]
You're probably using an older version of mypy. I think the bug you're hitting was fixed in 1.5, so if you upgrade, it should go away.
